### ISSUE TITLE
Speed-up common beam operations

### ISSUE THREAD

**Issue by e-koch** [radio-astro-tools/radio-beam#86]
For a beam table with 1178 channels:

```
beams.common_beam()                                                                                                              
3.64 s ± 427 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Caching the common beam will avoid recomputing but for really large cubes, this is getting pretty slow.

It may also be N times slower when the approximate methods used to calculate the common beam needs to be incrementally increased by `epsilon` (see #82). Currently that's needed to avoid the user needing to manually choose kwargs.

**Comment by e-koch** [comment]
* [ ] Move deconvolution, analytic stuff to cython
* [ ] Standardize units on input then drop until the end

**Comment by keflavich** [comment]
```
Timer unit: 1e-06 s

Total time: 4.09146 s
File: /home/adam/anaconda3/lib/python3.7/site-packages/radio_beam/beam.py
Function: deconvolve at line 359

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   359                                               def deconvolve(self, other, failure_returns_pointlike=False):
   360                                                   """
   361                                                   Deconvolve a beam from another
   362
   363                                                   Parameters
   364                                                   ----------
   365                                                   other : `Beam`
   366                                                       The beam to deconvolve from this beam
   367                                                   failure_returns_pointlike : bool
   368                                                       Option to return a pointlike beam (i.e., one with major=minor=0) if
   369                                                       the second beam is larger than the first.  Otherwise, a ValueError
   370                                                       will be raised
   371
   372                                                   Returns
   373                                                   -------
   374                                                   new_beam : `Beam`
   375                                                       The convolved Beam
   376
   377                                                   Raises
   378                                                   ------
   379                                                   failure : ValueError
   380                                                       If the second beam is larger than the first, the default behavior
   381                                                       is to raise an exception.  This can be overridden with
   382                                                       failure_returns_pointlike
   383                                                   """
   384
   385                                                   new_major, new_minor, new_pa = \
   386      1000        713.0      0.7      0.0              deconvolve(self, other,
   387      1000    3605000.0   3605.0     88.1                         failure_returns_pointlike=failure_returns_pointlike)
   388      1000     485750.0    485.8     11.9          return Beam(major=new_major, minor=new_minor, pa=new_pa)

Total time: 5.32446 s
File: /home/adam/anaconda3/lib/python3.7/site-packages/radio_beam/commonbeam.py
Function: fits_in_largest at line 331

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   331                                           def fits_in_largest(beams, large_beam=None):
   332                                               '''
   333                                               Test if all beams can be deconvolved by the largest beam
   334                                               '''
   335
   336         1          1.0      1.0      0.0      if large_beam is None:
   337                                                   large_beam = beams.largest_beam()
   338
   339      1001     435019.0    434.6      8.2      for beam in beams:
   340      1000     505112.0    505.1      9.5          if large_beam == beam:
   341                                                       continue
   342      1000       1005.0      1.0      0.0          deconv_beam = large_beam.deconvolve(beam,
   343      1000    4098846.0   4098.8     77.0                                              failure_returns_pointlike=True)
   344
   345      1000     284479.0    284.5      5.3          if not deconv_beam.isfinite:
   346                                                       return False
   347
   348         1          0.0      0.0      0.0      return True

Total time: 6.43936 s
File: /home/adam/anaconda3/lib/python3.7/site-packages/radio_beam/commonbeam.py
Function: common_manybeams_mve at line 480

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   480                                           def common_manybeams_mve(beams, tolerance=1e-4, nsamps=200, epsilon=5e-4):
   481                                               '''
   482                                               Calculate a common beam size using the Khachiyan Algorithm to find the
   483                                               minimum enclosing ellipse from all beam edges.
   484
   485                                               Parameters
   486                                               ----------
   487                                               beams : `~radio_beam.Beams`
   488                                                   Beams object.
   489                                               tolerance : float, optional
   490                                                   Allowed error range in the Khachiyan Algorithm. Decreasing the
   491                                                   tolerance by an order of magnitude requires an order of magnitude
   492                                                   more iterations to converge.
   493                                               nsamps : int, optional
   494                                                   Number of edge points to sample from each beam.
   495                                               epsilon : float, optional
   496                                                   Increase the radii of each beam by a factor of 1 + epsilon to ensure
   497                                                   the common beam can marginally be deconvolved for all beams. Small
   498                                                   deviations result from the finite sampling of points and the choice
   499                                                   of the tolerance.
   500
   501                                               Returns
   502                                               -------
   503                                               com_beam : `~radio_beam.Beam`
   504                                                   The common beam for all beams in the set.
   505                                               '''
   506
   507         1          2.0      2.0      0.0      if not HAS_SCIPY:
   508                                                   raise ImportError("common_manybeams_mve requires scipy.optimize.")
   509
   510         1          1.0      1.0      0.0      pts = []
   511
   512      1001     431729.0    431.3      6.7      for beam in beams:
   513      1000     181611.0    181.6      2.8          pts.append(ellipse_edges(beam, nsamps, epsilon=epsilon))
   514
   515         1       3226.0   3226.0      0.1      all_pts = np.hstack(pts).T
   516
   517                                               # Now find the outer edges of the convex hull.
   518         1      28304.0  28304.0      0.4      hull = ConvexHull(all_pts)
   519         1         26.0     26.0      0.0      edge_pts = all_pts[hull.vertices]
   520
   521                                               center, radii, rotation = \
   522         1     464316.0 464316.0      7.2          getMinVolEllipse(edge_pts, tolerance=tolerance)
   523
   524                                               # The rotation matrix is coming out as:
   525                                               # ((sin theta, cos theta)
   526                                               #  (cos theta, - sin theta))
   527         1         87.0     87.0      0.0      pa = np.arctan2(- rotation[0, 0], rotation[1, 0]) * u.rad
   528
   529         1          7.0      7.0      0.0      if pa.value == -np.pi or pa.value == np.pi:
   530                                                   pa = 0.0 * u.rad
   531
   532         1         81.0     81.0      0.0      com_beam = Beam(major=radii.max() * u.deg, minor=radii.min() * u.deg,
   533         1        668.0    668.0      0.0                      pa=pa)
   534
   535         1    5329305.0 5329305.0     82.8      if not fits_in_largest(beams, com_beam):
   536                                                   raise BeamError("Could not find common beam to deconvolve all beams.")
   537
   538         1          0.0      0.0      0.0      return com_beam

Total time: 3.56068 s
File: /home/adam/anaconda3/lib/python3.7/site-packages/radio_beam/utils.py
Function: deconvolve at line 19

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    19                                           def deconvolve(beam, other, failure_returns_pointlike=False):
    20                                               """
    21                                               Deconvolve a beam from another
    22
    23                                               Parameters
    24                                               ----------
    25                                               beam : `Beam`
    26                                                   The defined beam.
    27                                               other : `Beam`
    28                                                   The beam to deconvolve from this beam
    29                                               failure_returns_pointlike : bool
    30                                                   Option to return a pointlike beam (i.e., one with major=minor=0) if
    31                                                   the second beam is larger than the first.  Otherwise, a ValueError
    32                                                   will be raised
    33
    34                                               Returns
    35                                               -------
    36                                               new_beam : `Beam`
    37                                                   The convolved Beam
    38
    39                                               Raises
    40                                               ------
    41                                               failure : ValueError
    42                                                   If the second beam is larger than the first, the default behavior
    43                                                   is to raise an exception.  This can be overridden with
    44                                                   failure_returns_pointlike
    45                                               """
    46
    47                                               # blame: https://github.com/pkgw/carma-miriad/blob/CVSHEAD/src/subs/gaupar.for
    48                                               # (githup checkin of MIRIAD, code by Sault)
    49
    50                                               # rename to shorter variables for readability
    51      1000       2141.0      2.1      0.1      maj1, min1, pa1 = beam.major, beam.minor, beam.pa
    52      1000       1661.0      1.7      0.0      maj2, min2, pa2 = other.major, other.minor, other.pa
    53      1000       1114.0      1.1      0.0      cos, sin = np.cos, np.sin
    54
    55                                               alpha = ((maj1 * cos(pa1))**2 +
    56                                                        (min1 * sin(pa1))**2 -
    57      1000     454692.0    454.7     12.8               (maj2 * cos(pa2))**2 -
    58      1000     168255.0    168.3      4.7               (min2 * sin(pa2))**2)
    59
    60                                               beta = ((maj1 * sin(pa1))**2 +
    61                                                       (min1 * cos(pa1))**2 -
    62      1000     441390.0    441.4     12.4              (maj2 * sin(pa2))**2 -
    63      1000     166911.0    166.9      4.7              (min2 * cos(pa2))**2)
    64
    65      1000     273054.0    273.1      7.7      gamma = 2 * ((min1**2 - maj1**2) * sin(pa1) * cos(pa1) -
    66      1000     350433.0    350.4      9.8                   (min2**2 - maj2**2) * sin(pa2) * cos(pa2))
    67
    68      1000      33073.0     33.1      0.9      s = alpha + beta
    69      1000     246948.0    246.9      6.9      t = np.sqrt((alpha - beta)**2 + gamma**2)
    70
    71                                               # identify the smallest resolution
    72      1000      22180.0     22.2      0.6      axes = np.array([maj1.to(u.deg).value,
    73      1000      17623.0     17.6      0.5                       min1.to(u.deg).value,
    74      1000      26681.0     26.7      0.7                       maj2.to(u.deg).value,
    75      1000      53398.0     53.4      1.5                       min2.to(u.deg).value]) * u.deg
    76      1000      51738.0     51.7      1.5      limit = np.min(axes)
    77      1000      94765.0     94.8      2.7      limit = 0.1 * limit * limit
    78
    79                                               # Deal with floating point issues
    80      1000      29951.0     30.0      0.8      atol_t = np.finfo(t.dtype).eps * t.unit
    81
    82                                               # To deconvolve, the beam must satisfy:
    83                                               # alpha < 0
    84      1000      75541.0     75.5      2.1      alpha_cond = alpha.to(u.arcsec**2).value + np.finfo(alpha.dtype).eps < 0
    85                                               # beta < 0
    86      1000      68164.0     68.2      1.9      beta_cond = beta.to(u.arcsec**2).value + np.finfo(beta.dtype).eps < 0
    87                                               # s < t
    88      1000      99426.0     99.4      2.8      st_cond = s < t + atol_t
    89
    90      1000       1267.0      1.3      0.0      if alpha_cond or beta_cond or st_cond:
    91                                                   if failure_returns_pointlike:
    92                                                       return 0 * maj1.unit, 0 * min1.unit, 0 * pa1.unit
    93                                                   else:
    94                                                       raise ValueError("Beam could not be deconvolved")
    95                                               else:
    96      1000     115757.0    115.8      3.3          new_major = np.sqrt(0.5 * (s + t))
    97      1000     112004.0    112.0      3.1          new_minor = np.sqrt(0.5 * (s - t))
    98
    99                                                   # absolute tolerance needs to be <<1 microarcsec
   100      1000     388332.0    388.3     10.9          if np.isclose(((abs(gamma) + abs(alpha - beta))**0.5).to(u.arcsec).value, 1e-7):
   101                                                       new_pa = 0.0
   102                                                   else:
   103      1000     138435.0    138.4      3.9              new_pa = 0.5 * np.arctan2(-1. * gamma, alpha - beta)
   104
   105                                               # In the limiting case, the axes can be zero to within precision
   106                                               # Add the precision level onto each axis so a deconvolvable beam
   107                                               # is always has beam.isfinite == True
   108      1000      66491.0     66.5      1.9      new_major += np.finfo(new_major.dtype).eps * new_major.unit
   109      1000      58205.0     58.2      1.6      new_minor += np.finfo(new_minor.dtype).eps * new_minor.unit
   110
   111      1000       1049.0      1.0      0.0      return new_major, new_minor, new_pa


Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    36                                               def __new__(cls, major=None, minor=None, pa=None, area=None,
    37                                                           default_unit=u.arcsec, meta=None):
    38                                                   """
    39                                                   Create a new Gaussian beam
    40
    41                                                   Parameters
    42                                                   ----------
    43                                                   major : :class:`~astropy.units.Quantity` with angular equivalency
    44                                                       The FWHM major axis
    45                                                   minor : :class:`~astropy.units.Quantity` with angular equivalency
    46                                                       The FWHM minor axis
    47                                                   pa : :class:`~astropy.units.Quantity` with angular equivalency
    48                                                       The beam position angle
    49                                                   area : :class:`~astropy.units.Quantity` with steradian equivalency
    50                                                       The area of the beam.  This is an alternative to specifying the
    51                                                       major/minor/PA, and will create those values assuming a circular
    52                                                       Gaussian beam.
    53                                                   default_unit : :class:`~astropy.units.Unit`
    54                                                       The unit to impose on major, minor if they are specified as floats
    55                                                   """
    56
    57                                                   # improve to some kwargs magic later
    58
    59                                                   # error checking
    60
    61                                                   # ... given an area make a round beam assuming it is Gaussian
    62      3001       2783.0      0.9      0.2          if area is not None:
    63                                                       if major is not None:
    64                                                           raise ValueError("Can only specify one of {major,minor,pa} "
    65                                                                            "and {area}")
    66                                                       rad = np.sqrt(area/(2*np.pi)) * u.deg
    67                                                       major = rad * SIGMA_TO_FWHM
    68                                                       minor = rad * SIGMA_TO_FWHM
    69                                                       pa = 0.0 * u.deg
    70
    71                                                   # give specified values priority
    72      3001       2213.0      0.7      0.2          if major is not None:
    73      3001     242322.0     80.7     18.0              if u.deg.is_equivalent(major):
    74      3001       2732.0      0.9      0.2                  major = major
    75                                                       else:
    76                                                           warnings.warn("Assuming major axis has been specified in degrees")
    77                                                           major = major * u.deg
    78      3001       2267.0      0.8      0.2          if minor is not None:
    79      3001     206090.0     68.7     15.3              if u.deg.is_equivalent(minor):
    80      3001       2625.0      0.9      0.2                  minor = minor
    81                                                       else:
    82                                                           warnings.warn("Assuming minor axis has been specified in degrees")
    83                                                           minor = minor * u.deg
    84      3001       2179.0      0.7      0.2          if pa is not None:
    85      3001     171679.0     57.2     12.8              if u.deg.is_equivalent(pa):
    86      3001       2536.0      0.8      0.2                  pa = pa
    87                                                       else:
    88                                                           warnings.warn("Assuming position angle has been specified in degrees")
    89                                                           pa = pa * u.deg
    90                                                   else:
    91                                                       pa = 0.0 * u.deg
    92
    93                                                   # some sensible defaults
    94      3001       2258.0      0.8      0.2          if minor is None:
    95                                                       minor = major
    96
    97      3001     158748.0     52.9     11.8          if minor > major:
    98                                                       raise ValueError("Minor axis greater than major axis.")
    99
   100      3001     528928.0    176.3     39.3          self = super(Beam, cls).__new__(cls, _to_area(major,minor).value, u.sr)
   101      3001       3136.0      1.0      0.2          self._major = major
   102      3001       2281.0      0.8      0.2          self._minor = minor
   103      3001       2182.0      0.7      0.2          self._pa = pa
   104      3001       2154.0      0.7      0.2          self.default_unit = default_unit
   105
   106      3001       2204.0      0.7      0.2          if meta is None:
   107      1001        818.0      0.8      0.1              self.meta = {}
   108      2000       1758.0      0.9      0.1          elif isinstance(meta, dict):
   109      2000       1515.0      0.8      0.1              self.meta = meta
   110                                                   else:
   111                                                       raise TypeError("metadata must be a dictionary")
   112
   113      3001       2010.0      0.7      0.1          return self

```

**Comment by keflavich** [comment]
Looks like removing quantity-related operations in the internal functions could save a ton of time

**Comment by keflavich** [comment]
cmd for that timing:
```
import numpy as np
from radio_beam import commonbeam
import radio_beam
from astropy import units as u

y = np.random.randn(1000)*0.01 + 1
x = np.random.randn(1000)*0.01 + 1
bpa = np.arcsin(np.random.random(1000))
maj = np.max([x,y], axis=0)
min = np.min([x,y], axis=0)
beams = radio_beam.Beams(major=maj*u.arcsec, minor=min*u.arcsec, pa=bpa*u.deg)
%lprun -f radio_beam.Beam.__new__ -f radio_beam.utils.deconvolve -f radio_beam.Beam.deconvolve -f commonbeam.fits_in_largest  -f commonbeam.common_manybeams_mve  commonbeam.common_manybeams_mve(beams)
```

**Comment by e-koch** [comment]
With #86, the same timing example at the beginning is now:

```
681 ms ± 72.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
so factor of 5 improvement

**Comment by e-koch** [comment]
```
%load_ext line_profiler

import numpy as np
from radio_beam import commonbeam
import radio_beam
from astropy import units as u

y = np.random.randn(1000)*0.01 + 1
x = np.random.randn(1000)*0.01 + 1
bpa = np.arcsin(np.random.random(1000))
maj = np.max([x,y], axis=0)
min = np.min([x,y], axis=0)
beams = radio_beam.Beams(major=maj*u.arcsec, minor=min*u.arcsec, pa=bpa*u.deg)
%lprun -f radio_beam.Beam.__new__ -f radio_beam.utils.deconvolve_opt -f commonbeam.fits_in_largest  -f commonbeam.common_manybeams_mve  commonbeam.common_manybeams_mve(beams)
```

```
Total time: 0.462345 s
File: /home/ekoch/ownCloud/code_development/radio_astro_tools/radio_beam/radio_beam/beam.py
Function: __new__ at line 36

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    36                                               def __new__(cls, major=None, minor=None, pa=None, area=None,
    37                                                           default_unit=u.arcsec, meta=None):
    38                                                   """
    39                                                   Create a new Gaussian beam
    40                                           
    41                                                   Parameters
    42                                                   ----------
    43                                                   major : :class:`~astropy.units.Quantity` with angular equivalency
    44                                                       The FWHM major axis
    45                                                   minor : :class:`~astropy.units.Quantity` with angular equivalency
    46                                                       The FWHM minor axis
    47                                                   pa : :class:`~astropy.units.Quantity` with angular equivalency
    48                                                       The beam position angle
    49                                                   area : :class:`~astropy.units.Quantity` with steradian equivalency
    50                                                       The area of the beam.  This is an alternative to specifying the
    51                                                       major/minor/PA, and will create those values assuming a circular
    52                                                       Gaussian beam.
    53                                                   default_unit : :class:`~astropy.units.Unit`
    54                                                       The unit to impose on major, minor if they are specified as floats
    55                                                   """
    56                                           
    57                                                   # improve to some kwargs magic later
    58                                           
    59                                                   # error checking
    60                                           
    61                                                   # ... given an area make a round beam assuming it is Gaussian
    62      1001        868.0      0.9      0.2          if area is not None:
    63                                                       if major is not None:
    64                                                           raise ValueError("Can only specify one of {major,minor,pa} "
    65                                                                            "and {area}")
    66                                                       rad = np.sqrt(area/(2*np.pi)) * u.deg
    67                                                       major = rad * SIGMA_TO_FWHM
    68                                                       minor = rad * SIGMA_TO_FWHM
    69                                                       pa = 0.0 * u.deg
    70                                           
    71                                                   # give specified values priority
    72      1001        617.0      0.6      0.1          if major is not None:
    73      1001      86964.0     86.9     18.8              if u.deg.is_equivalent(major):
    74      1001        805.0      0.8      0.2                  major = major
    75                                                       else:
    76                                                           warnings.warn("Assuming major axis has been specified in degrees")
    77                                                           major = major * u.deg
    78      1001        557.0      0.6      0.1          if minor is not None:
    79      1001      75256.0     75.2     16.3              if u.deg.is_equivalent(minor):
    80      1001        758.0      0.8      0.2                  minor = minor
    81                                                       else:
    82                                                           warnings.warn("Assuming minor axis has been specified in degrees")
    83                                                           minor = minor * u.deg
    84      1001        553.0      0.6      0.1          if pa is not None:
    85      1001      71299.0     71.2     15.4              if u.deg.is_equivalent(pa):
    86      1001        764.0      0.8      0.2                  pa = pa
    87                                                       else:
    88                                                           warnings.warn("Assuming position angle has been specified in degrees")
    89                                                           pa = pa * u.deg
    90                                                   else:
    91                                                       pa = 0.0 * u.deg
    92                                           
    93                                                   # some sensible defaults
    94      1001        591.0      0.6      0.1          if minor is None:
    95                                                       minor = major
    96                                           
    97      1001      28028.0     28.0      6.1          if minor > major:
    98                                                       raise ValueError("Minor axis greater than major axis.")
    99                                           
   100      1001     190323.0    190.1     41.2          self = super(Beam, cls).__new__(cls, _to_area(major,minor).value, u.sr)
   101      1001        908.0      0.9      0.2          self._major = major
   102      1001        602.0      0.6      0.1          self._minor = minor
   103      1001        529.0      0.5      0.1          self._pa = pa
   104      1001        571.0      0.6      0.1          self.default_unit = default_unit
   105                                           
   106      1001        586.0      0.6      0.1          if meta is None:
   107         1          1.0      1.0      0.0              self.meta = {}
   108      1000        720.0      0.7      0.2          elif isinstance(meta, dict):
   109      1000        573.0      0.6      0.1              self.meta = meta
   110                                                   else:
   111                                                       raise TypeError("metadata must be a dictionary")
   112                                           
   113      1001        472.0      0.5      0.1          return self

Total time: 0.21595 s
File: /home/ekoch/ownCloud/code_development/radio_astro_tools/radio_beam/radio_beam/commonbeam.py
Function: fits_in_largest at line 331

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   331                                           def fits_in_largest(beams, large_beam=None):
   332                                               '''
   333                                               Test if all beams can be deconvolved by the largest beam
   334                                               '''
   335                                           
   336         2          2.0      1.0      0.0      if large_beam is None:
   337                                                   large_beam = beams.largest_beam()
   338                                           
   339         2        210.0    105.0      0.1      large_hdr_keywords = large_beam.to_header_keywords()
   340                                           
   341         2         78.0     39.0      0.0      majors = beams.major.to(u.deg).value
   342         2         69.0     34.5      0.0      minors = beams.minor.to(u.deg).value
   343         2         46.0     23.0      0.0      pas = beams.pa.to(u.deg).value
   344                                           
   345      2002       2968.0      1.5      1.4      for major, minor, pa in zip(majors, minors, pas):
   346                                           
   347      2000       2879.0      1.4      1.3          equal = abs(large_hdr_keywords['BMAJ'] - major) < 1e-12
   348      2000       1605.0      0.8      0.7          equal = equal and (abs(large_hdr_keywords['BMIN'] - minor) < 1e-12)
   349                                           
   350                                                   # Check if the beam is circular
   351      2000       2182.0      1.1      1.0          iscircular = (major - minor) / minor < 1e-6
   352                                           
   353                                                   # position angle only matters if the beam is asymmetric
   354      2000       1527.0      0.8      0.7          if not iscircular:
   355      2000       1444.0      0.7      0.7              equal = equal and (abs(((large_hdr_keywords['BPA'] % np.pi) - (pa % np.pi))) < 1e-12)
   356                                           
   357      2000       1476.0      0.7      0.7          if equal:
   358                                                       continue
   359                                           
   360      2000       1479.0      0.7      0.7          out = deconvolve_opt(large_hdr_keywords,
   361      2000       1423.0      0.7      0.7                               {'BMAJ': major,
   362      2000       1418.0      0.7      0.7                                'BMIN': minor,
   363      2000       1667.0      0.8      0.8                                'BPA': pa},
   364      2000     166341.0     83.2     77.0                               failure_returns_pointlike=True)
   365                                           
   366      2000      29136.0     14.6     13.5          if np.any([ax == 0. for ax in out[:2]]):
   367                                                       return False
   368                                           
   369         2          0.0      0.0      0.0      return True

Total time: 1.50287 s
File: /home/ekoch/ownCloud/code_development/radio_astro_tools/radio_beam/radio_beam/commonbeam.py
Function: common_manybeams_mve at line 501

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   501                                           def common_manybeams_mve(beams, tolerance=1e-4, nsamps=200,
   502                                                                    epsilon=5e-4,
   503                                                                    auto_increase_epsilon=True,
   504                                                                    max_epsilon=1e-3,
   505                                                                    max_iter=10):
   506                                               '''
   507                                               Calculate a common beam size using the Khachiyan Algorithm to find the
   508                                               minimum enclosing ellipse from all beam edges.
   509                                           
   510                                               Parameters
   511                                               ----------
   512                                               beams : `~radio_beam.Beams`
   513                                                   Beams object.
   514                                               tolerance : float, optional
   515                                                   Allowed error range in the Khachiyan Algorithm. Decreasing the
   516                                                   tolerance by an order of magnitude requires an order of magnitude
   517                                                   more iterations to converge.
   518                                               nsamps : int, optional
   519                                                   Number of edge points to sample from each beam.
   520                                               epsilon : float, optional
   521                                                   Increase the radii of each beam by a factor of 1 + epsilon to ensure
   522                                                   the common beam can marginally be deconvolved for all beams. Small
   523                                                   deviations result from the finite sampling of points and the choice
   524                                                   of the tolerance.
   525                                               auto_increase_epsilon : bool, optional
   526                                                   Re-run the algorithm when the solution cannot quite be deconvolved from
   527                                                   from all the beams. When `True`, epsilon is slightly increased with
   528                                                   each iteration until the common beam can be deconvolved from all beams.
   529                                                   Default is `True`.
   530                                               max_epsilon : float, optional
   531                                                   Maximum epsilon value that is acceptable. Reached with `max_iter`.
   532                                                   Default is `1e-3`.
   533                                               max_iter : int, optional
   534                                                   Maximum number of times to increase epsilon to try finding a valid
   535                                                   common beam solution.
   536                                           
   537                                               Returns
   538                                               -------
   539                                               com_beam : `~radio_beam.Beam`
   540                                                   The common beam for all beams in the set.
   541                                               '''
   542                                           
   543         1          2.0      2.0      0.0      if not HAS_SCIPY:
   544                                                   raise ImportError("common_manybeams_mve requires scipy.optimize.")
   545                                           
   546         1          0.0      0.0      0.0      step = 1
   547                                           
   548         1          1.0      1.0      0.0      while True:
   549         1          1.0      1.0      0.0          pts = []
   550                                                                                  
   551      1001     508637.0    508.1     33.8          for beam in beams:
   552      1000     188344.0    188.3     12.5              pts.append(ellipse_edges(beam, nsamps, epsilon=epsilon))
   553                                           
   554         1       2309.0   2309.0      0.2          all_pts = np.hstack(pts).T
   555                                           
   556                                                   # Now find the outer edges of the convex hull.
   557         1      29800.0  29800.0      2.0          hull = ConvexHull(all_pts)
   558         1         17.0     17.0      0.0          edge_pts = all_pts[hull.vertices]
   559                                           
   560                                                   center, radii, rotation = \
   561         1     532983.0 532983.0     35.5              getMinVolEllipse(edge_pts, tolerance=tolerance)
   562                                           
   563                                                   # The rotation matrix is coming out as:
   564                                                   # ((sin theta, cos theta)
   565                                                   #  (cos theta, - sin theta))
   566         1         82.0     82.0      0.0          pa = np.arctan2(- rotation[0, 0], rotation[1, 0]) * u.rad
   567                                           
   568         1         10.0     10.0      0.0          if pa.value == -np.pi or pa.value == np.pi:
   569                                                       pa = 0.0 * u.rad
   570                                           
   571         1         95.0     95.0      0.0          com_beam = Beam(major=radii.max() * u.deg, minor=radii.min() * u.deg,
   572         1        676.0    676.0      0.0                          pa=pa)
   573                                           
   574                                                   # If common beam is just slightly smaller than one of the beams,
   575                                                   # we increase epsilon to encourage a solution marginally larger
   576                                                   # so all beams can be convolved.
   577         1          1.0      1.0      0.0          if auto_increase_epsilon:
   578         1     124975.0 124975.0      8.3              if not fits_in_largest(beams, com_beam):
   579                                                           # Increase epsilon and run again
   580                                                           epsilon += (step + 1) * (max_epsilon - epsilon) / max_iter
   581                                                           step += 1
   582                                           
   583                                                           if step == max_iter + 1:
   584                                                               raise BeamError("Could not increase epsilon to find"
   585                                                                               " common beam.")
   586                                           
   587                                                           continue
   588                                                       else:
   589         1          1.0      1.0      0.0                  break
   590                                                   else:
   591                                                       break
   592                                           
   593         1     114935.0 114935.0      7.6      if not fits_in_largest(beams, com_beam):
   594                                                   raise BeamError("Could not find common beam to deconvolve all beams.")
   595                                           
   596         1          0.0      0.0      0.0      return com_beam

Total time: 0.096836 s
File: /home/ekoch/ownCloud/code_development/radio_astro_tools/radio_beam/radio_beam/utils.py
Function: deconvolve_opt at line 22

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    22                                           def deconvolve_opt(beamprops1, beamprops2, failure_returns_pointlike=False):
    23                                               """
    24                                               An optimized, non-Quantity version of beam deconvolution.
    25                                           
    26                                               Because no unit conversions are handled, the inputs MUST be in degrees for the major, minor,
    27                                               and position angle.
    28                                           
    29                                               Parameters
    30                                               ----------
    31                                               beamprops1: dict
    32                                                   Dictionary with keys 'BMAJ', 'BMIN', and 'BPA' for the beam to deconvolve from. Can be
    33                                                   produced with `~radio_beam.Beam.to_fits_keywords`.
    34                                               beamprops2: dict
    35                                                   Same as `beamprops1` for the second beam.
    36                                               failure_returns_pointlike : bool, optional
    37                                                   Return a point beam (zero area) when deconvolution fails. If `False`,
    38                                                   this will instead raise a `~radio_beam.utils.BeamError` when deconvolution fails.
    39                                           
    40                                           
    41                                               Returns
    42                                               -------
    43                                               new_major : float
    44                                                   Deconvolved major FWHM.
    45                                               new_minor : float
    46                                                   Deconvolved minor FWHM.
    47                                               new_pa : float
    48                                                   Deconvolved position angle.
    49                                           
    50                                               """
    51                                           
    52      2000       2042.0      1.0      2.1      maj1 = beamprops1['BMAJ']
    53      2000       1835.0      0.9      1.9      min1 = beamprops1['BMIN']
    54      2000       2189.0      1.1      2.3      pa1 = beamprops1['BPA'] * DEG2RAD
    55                                           
    56      2000       1828.0      0.9      1.9      maj2 = beamprops2['BMAJ']
    57      2000       1788.0      0.9      1.8      min2 = beamprops2['BMIN']
    58      2000       2060.0      1.0      2.1      pa2 = beamprops2['BPA'] * DEG2RAD
    59                                           
    60                                               alpha = ((maj1 * math.cos(pa1))**2 +
    61                                                        (min1 * math.sin(pa1))**2 -
    62      2000       7063.0      3.5      7.3               (maj2 * math.cos(pa2))**2 -
    63      2000       3351.0      1.7      3.5               (min2 * math.sin(pa2))**2)
    64                                           
    65                                               beta = ((maj1 * math.sin(pa1))**2 +
    66                                                       (min1 * math.cos(pa1))**2 -
    67      2000       5676.0      2.8      5.9              (maj2 * math.sin(pa2))**2 -
    68      2000       3241.0      1.6      3.3              (min2 * math.cos(pa2))**2)
    69                                           
    70      2000       4524.0      2.3      4.7      gamma = 2 * ((min1**2 - maj1**2) * math.sin(pa1) * math.cos(pa1) -
    71      2000       5382.0      2.7      5.6                   (min2**2 - maj2**2) * math.sin(pa2) * math.cos(pa2))
    72                                           
    73      2000       1980.0      1.0      2.0      s = alpha + beta
    74      2000       3991.0      2.0      4.1      t = math.sqrt((alpha - beta)**2 + gamma**2)
    75                                           
    76                                               # Deal with floating point issues
    77                                               # This matches the arcsec**2 check for deconvolve below
    78                                               # Difference is we keep things in deg^2 here
    79      2000       6906.0      3.5      7.1      atol_t = np.finfo(np.float64).eps / 3600.**2
    80                                           
    81                                               # To deconvolve, the beam must satisfy:
    82                                               # alpha < 0
    83      2000       6109.0      3.1      6.3      alpha_cond = alpha + np.finfo(np.float64).eps < 0
    84                                           
    85                                               # beta < 0
    86      2000       5817.0      2.9      6.0      beta_cond = beta + np.finfo(np.float64).eps < 0
    87                                               # s < t
    88      2000       2215.0      1.1      2.3      st_cond = s < t + atol_t
    89                                           
    90      2000       2027.0      1.0      2.1      if alpha_cond or beta_cond or st_cond:
    91                                                   if failure_returns_pointlike:
    92                                                       return 0., 0., 0.
    93                                                   else:
    94                                                       raise BeamError("Beam could not be deconvolved")
    95                                               else:
    96      2000       2929.0      1.5      3.0          new_major = math.sqrt(0.5 * (s + t))
    97      2000       2658.0      1.3      2.7          new_minor = math.sqrt(0.5 * (s - t))
    98                                           
    99                                                   # absolute tolerance needs to be <<1 microarcsec
   100      2000       1795.0      0.9      1.9          atol = 1e-7 / 3600.
   101      2000       3399.0      1.7      3.5          if (math.sqrt(abs(gamma) + abs(alpha - beta))) < atol:
   102                                                       new_pa = 0.0
   103                                                   else:
   104      2000       3285.0      1.6      3.4              new_pa = 0.5 * math.atan2(-1. * gamma, alpha - beta)
   105                                           
   106                                               # In the limiting case, the axes can be zero to within precision
   107                                               # Add the precision level onto each axis so a deconvolvable beam
   108                                               # is always has beam.isfinite == True
   109      2000       5649.0      2.8      5.8      new_major += np.finfo(np.float64).eps
   110      2000       5201.0      2.6      5.4      new_minor += np.finfo(np.float64).eps
   111                                           
   112      2000       1896.0      0.9      2.0      return new_major, new_minor, new_pa

**Comment by e-koch** [comment]
Addressed in #87
